File: src/core/update.ts
/**
 * Self-Update Module
 * Handles version checking and binary updates using GitHub releases via gh CLI
 */

import { $ } from "bun";
import { logger } from "./logger";
import { getPlatform, getArch, getHomeDir, commandExists } from "./os";
import { loadConfig } from "./config";
import { createHookContext, runHook } from "./hooks";
import type { UpdateState, GitHubRelease, InstallOptions, UpdateHookContext } from "../types";
import pkg from "../../package.json";

const REPO = "alexcatdad/paw";
const STATE_FILE = `${getHomeDir()}/.paw-state.json`;
const CHECK_INTERVAL_MS = 24 * 60 * 60 * 1000; // 24 hours

/**
 * Load update state from disk
 */
export async function loadUpdateState(): Promise<UpdateState | null> {
  try {
    const file = Bun.file(STATE_FILE);
    if (await file.exists()) {
      return await file.json();
    }
  } catch {
    // State file doesn't exist or is invalid
  }
  return null;
}

/**
 * Save update state to disk
 */
export async function saveUpdateState(state: UpdateState): Promise<void> {
  await Bun.write(STATE_FILE, JSON.stringify(state, null, 2));
}

/**
 * Check if we should skip the version check (rate limiting)
 */
export async function shouldSkipCheck(): Promise<boolean> {
  const state = await loadUpdateState();
  if (!state) return false;

  const lastCheck = new Date(state.lastCheck).getTime();
  const now = Date.now();
  return now - lastCheck < CHECK_INTERVAL_MS;
}

/**
 * Get the latest release from GitHub using gh CLI
 */
export async function getLatestRelease(): Promise<GitHubRelease | null> {
  // Check if gh is available
  if (!(await commandExists("gh"))) {
    logger.warn("gh CLI not found. Install with: brew install gh");
    return null;
  }

  try {
    const result = await $`gh release view latest --repo ${REPO} --json tagName,assets`.quiet().nothrow();

    if (result.exitCode !== 0) {
      // Could be offline, no releases, or not authenticated
      const stderr = result.stderr.toString();
      if (stderr.includes("release not found")) {
        logger.debug?.("No releases found");
      } else if (stderr.includes("authentication")) {
        logger.warn("gh not authenticated. Run: gh auth login");
      }
      return null;
    }

    const data = JSON.parse(result.text());
    return {
      tag_name: data.tagName,
      assets: data.assets.map((a: { name: string; url: string }) => ({
        name: a.name,
        browser_download_url: a.url,
      })),
    };
  } catch (error) {
    logger.debug?.(`Failed to fetch release: ${error}`);
    return null;
  }
}

/**
 * Compare two semantic versions
 * Returns: -1 if a < b, 0 if a == b, 1 if a > b
 */
export function compareVersions(a: string, b: string): number {
  // Strip leading 'v' if present
  const cleanA = a.replace(/^v/, "");
  const cleanB = b.replace(/^v/, "");

  const partsA = cleanA.split(".").map(Number);
  const partsB = cleanB.split(".").map(Number);

  for (let i = 0; i < Math.max(partsA.length, partsB.length); i++) {
    const numA = partsA[i] ?? 0;
    const numB = partsB[i] ?? 0;

    if (numA < numB) return -1;
    if (numA > numB) return 1;
  }

  return 0;
}

/**
 * Get the expected binary name for the current platform
 */
export function getBinaryName(): string {
  const platform = getPlatform();
  const arch = getArch();
  return `paw-${platform}-${arch}`;
}

/**
 * Get the path where paw is installed
 */
export async function getPawBinaryPath(): Promise<string | null> {
  try {
    const result = await $`which paw`.quiet().nothrow();
    if (result.exitCode === 0) {
      return result.text().trim();
    }
  } catch {
    // paw not in PATH
  }

  // Check default location
  const defaultPath = `${getHomeDir()}/.local/bin/paw`;
  const file = Bun.file(defaultPath);
  if (await file.exists()) {
    return defaultPath;
  }

  return null;
}

/**
 * Download the binary for the current platform with optional verification
 */
export async function downloadBinary(
  targetDir: string,
  options: InstallOptions & { skipVerify?: boolean }
): Promise<string | null> {
  const binaryName = getBinaryName();

  if (options.dryRun) {
    logger.info(`Would download: ${binaryName} to ${targetDir}`);
    return `${targetDir}/${binaryName}`;
  }

  try {
    // Check if gh supports --verify-attestation
    const helpResult = await $`gh release download --help`.quiet().nothrow();
    const supportsVerify = helpResult.text().includes("verify-attestation");

    if (supportsVerify && !options.skipVerify) {
      logger.info("Downloading with attestation verification...");
      const result = await $`gh release download latest --repo ${REPO} --pattern ${binaryName} --dir ${targetDir} --verify-attestation`.quiet().nothrow();

      if (result.exitCode !== 0) {
        const stderr = result.stderr.toString();

        if (stderr.includes("attestation")) {
          logger.error("Binary verification failed - attestation invalid or missing");
          logger.info("Use --skip-verify to bypass verification (not recommended)");
          return null;
        }

        logger.error(`Failed to download binary: ${stderr}`);
        return null;
      }

      logger.success("Binary attestation verified");
    } else {
      if (!options.skipVerify && !supportsVerify) {
        logger.warn("gh CLI doesn't support attestation verification. Consider updating gh.");
      }

      const result = await $`gh release download latest --repo ${REPO} --pattern ${binaryName} --dir ${targetDir}`.quiet().nothrow();

      if (result.exitCode !== 0) {
        logger.error(`Failed to download binary: ${result.stderr.toString()}`);
        return null;
      }
    }

    return `${targetDir}/${binaryName}`;
  } catch (error) {
    logger.error(`Download failed: ${error}`);
    return null;
  }
}

/**
 * Perform the self-update
 */
export async function performUpdate(options: InstallOptions & { skipVerify?: boolean }): Promise<boolean> {
  const currentVersion = pkg.version;
  const binaryPath = await getPawBinaryPath();

  if (!binaryPath) {
    logger.error("Could not find paw binary. Are you running from source?");
    logger.info("For source installs, update with: git pull && bun install");
    return false;
  }

  // Check if this is a compiled binary
  // In compiled Bun binaries, process.argv[0] is "bun" and process.execPath is the binary path
  // In dev mode (bun run), process.argv[0] is the full path to bun like /opt/homebrew/bin/bun
  const isCompiled = !binaryPath.endsWith(".ts") && (process.argv[0] === "bun" || !process.argv[0]?.includes("/"));
  if (!isCompiled && !options.force) {
    logger.warn("Running from source, not a compiled binary.");
    logger.info("To update: git pull && bun install");
    return false;
  }

  logger.info(`Current version: v${currentVersion}`);
  logger.info(`Binary location: ${binaryPath}`);

  // Fetch latest release
  const release = await getLatestRelease();
  if (!release) {
    logger.error("Could not fetch latest release");
    return false;
  }

  const latestVersion = release.tag_name.replace(/^v/, "");
  logger.info(`Latest version: v${latestVersion}`);

  // Compare versions
  const comparison = compareVersions(currentVersion, latestVersion);
  if (comparison >= 0) {
    logger.success("Already up to date!");
    await saveUpdateState({
      lastCheck: new Date().toISOString(),
      latestVersion,
      currentVersion,
    });
    return true;
  }

  logger.info(`Update available: v${currentVersion} → v${latestVersion}`);

  // Load config for hooks
  let config;
  try {
    config = await loadConfig();
  } catch {
    // Config might not exist, that's ok
  }

  // Run preUpdate hook
  if (config?.hooks?.preUpdate) {
    await runHook(config, "preUpdate", createHookContext(options.dryRun));
  }

  if (options.dryRun) {
    logger.info("Dry run - no changes made");
    return true;
  }

  // Check if we have the right binary in the release
  const binaryName = getBinaryName();
  const hasAsset = release.assets.some((a) => a.name === binaryName);
  if (!hasAsset) {
    logger.error(`No binary found for your platform: ${binaryName}`);
    logger.info(`Available: ${release.assets.map((a) => a.name).join(", ")}`);
    return false;
  }

  // Create temp directory for download
  const tmpDir = `/tmp/paw-update-${Date.now()}`;
  await $`mkdir -p ${tmpDir}`.quiet();

  try {
    // Download new binary
    logger.info("Downloading new binary...");
    const downloadedPath = await downloadBinary(tmpDir, options);
    if (!downloadedPath) {
      return false;
    }

    // Backup current binary
    const backupPath = `${binaryPath}.backup`;
    logger.info(`Backing up current binary to ${backupPath}`);
    await $`cp ${binaryPath} ${backupPath}`.quiet();

    // Install new binary
    logger.info("Installing new binary...");
    await $`mv ${downloadedPath} ${binaryPath}`.quiet();
    await $`chmod +x ${binaryPath}`.quiet();

    // Verify new binary works
    logger.info("Verifying installation...");
    const verifyResult = await $`${binaryPath} --version`.quiet().nothrow();

    if (verifyResult.exitCode !== 0) {
      // Restore backup
      logger.error("Verification failed, restoring backup...");
      await $`mv ${backupPath} ${binaryPath}`.quiet();
      return false;
    }

    // Remove backup
    await $`rm -f ${backupPath}`.quiet();

    // Update state
    await saveUpdateState({
      lastCheck: new Date().toISOString(),
      latestVersion,
      currentVersion: latestVersion,
    });

    // Run postUpdate hook
    if (config?.hooks?.postUpdate) {
      const updateContext: UpdateHookContext = {
        ...createHookContext(options.dryRun),
        previousVersion: currentVersion,
        newVersion: latestVersion,
      };
      await runHook(config, "postUpdate", updateContext);
    }

    logger.success(`Updated to v${latestVersion}!`);
    return true;
  } finally {
    // Cleanup temp directory
    await $`rm -rf ${tmpDir}`.quiet().nothrow();
  }
}

/**
 * Check for updates (used by sync command)
 * Returns the latest version if an update is available, null otherwise
 */
export async function checkForUpdate(options: { force?: boolean } = {}): Promise<string | null> {
  // Skip if we checked recently (unless forced)
  if (!options.force && (await shouldSkipCheck())) {
    return null;
  }

  const release = await getLatestRelease();
  if (!release) {
    return null;
  }

  const currentVersion = pkg.version;
  const latestVersion = release.tag_name.replace(/^v/, "");

  // Update state
  await saveUpdateState({
    lastCheck: new Date().toISOString(),
    latestVersion,
    currentVersion,
  });

  // Return latest version if update is available
  if (compareVersions(currentVersion, latestVersion) < 0) {
    return latestVersion;
  }

  return null;
}

File: internal/update/update.go
package update

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"runtime"
	"strings"
	"time"

	"github.com/alexcatdad/paw/internal/output"
	"github.com/alexcatdad/paw/internal/state"
)

const repoName = "alexcatdad/paw"
const checkInterval = 24 * time.Hour

const (
	ghReleaseCheckTimeout    = 30 * time.Second
	ghReleaseDownloadTimeout = 3 * time.Minute
)

type updateState struct {
	LastCheck      string `json:"lastCheck"`
	LatestVersion  string `json:"latestVersion"`
	CurrentVersion string `json:"currentVersion"`
}

type Options struct {
	DryRun     bool
	SkipVerify bool
	ForceCheck bool
}

func CheckForUpdate(currentVersion string, force bool) (string, error) {
	if !force {
		s, _ := loadState()
		if s != nil {
			last, err := time.Parse(time.RFC3339, s.LastCheck)
			if err == nil && time.Since(last) < checkInterval {
				if compareVersions(currentVersion, s.LatestVersion) < 0 {
					return s.LatestVersion, nil
				}
				return "", nil
			}
		}
	}
	if _, err := getRunner().LookPath("gh"); err != nil {
		return "", nil
	}

	ctx, cancel := context.WithTimeout(context.Background(), ghReleaseCheckTimeout)
	defer cancel()
	out, err := getRunner().OutputContext(ctx, "gh", "release", "view", "latest", "--repo", repoName, "--json", "tagName")
	if err != nil {
		return "", nil
	}
	var payload struct {
		TagName string `json:"tagName"`
	}
	if err := json.Unmarshal(out, &payload); err != nil {
		return "", err
	}
	latest := strings.TrimPrefix(strings.TrimSpace(payload.TagName), "v")
	_ = saveState(updateState{LastCheck: getClk().Now().UTC().Format(time.RFC3339), LatestVersion: latest, CurrentVersion: currentVersion})
	if compareVersions(currentVersion, latest) < 0 {
		return latest, nil
	}
	return "", nil
}

func Perform(currentVersion string, opts Options, logger *output.Logger) error {
	if _, err := getRunner().LookPath("gh"); err != nil {
		return fmt.Errorf("gh CLI not found")
	}
	latest, err := CheckForUpdate(currentVersion, opts.ForceCheck)
	if err != nil {
		return err
	}
	if latest == "" {
		logger.Success("Already up to date")
		return nil
	}
	binaryPath, err := getRunner().LookPath("paw")
	if err != nil {
		return fmt.Errorf("paw binary not found in PATH")
	}
	asset := fmt.Sprintf("paw-%s-%s", runtime.GOOS, mapArch(runtime.GOARCH))

	// Fix M6: use os.MkdirTemp for an unpredictable temp directory
	tmpDir, err := os.MkdirTemp("", "paw-update-*")
	if err != nil {
		return err
	}
	defer os.RemoveAll(tmpDir)

	if opts.DryRun {
		logger.DryRun(fmt.Sprintf("Would download %s and replace %s", asset, binaryPath))
		return nil
	}
	args := []string{"release", "download", "latest", "--repo", repoName, "--pattern", asset, "--dir", tmpDir}

	// Fix I2: warn explicitly when attestation verification is unavailable
	if opts.SkipVerify {
		fmt.Fprintf(os.Stderr, "WARNING: Binary verification skipped (--skip-verify flag).\n")
	} else {
		helpOut, _ := getRunner().Output("gh", "release", "download", "--help")
		if strings.Contains(string(helpOut), "verify-attestation") {
			args = append(args, "--verify-attestation")
		} else {
			fmt.Fprintf(os.Stderr, "WARNING: gh CLI does not support attestation verification. Update gh to enable binary verification. Proceeding without verification.\n")
		}
	}

	dlCtx, dlCancel := context.WithTimeout(context.Background(), ghReleaseDownloadTimeout)
	defer dlCancel()
	if out, err := getRunner().CombinedOutputContext(dlCtx, "gh", args...); err != nil {
		return fmt.Errorf("download failed: %s", strings.TrimSpace(string(out)))
	}

	downloaded := filepath.Join(tmpDir, asset)
	backup := binaryPath + ".backup"
	if err := copyFile(binaryPath, backup); err != nil {
		return err
	}
	if err := getFsys().Rename(downloaded, binaryPath); err != nil {
		_ = getFsys().Rename(backup, binaryPath)
		return err
	}
	if err := getFsys().Chmod(binaryPath, 0o755); err != nil {
		_ = getFsys().Rename(backup, binaryPath)
		return err
	}
	if out, err := getRunner().CombinedOutput(binaryPath, "--version"); err != nil {
		_ = getFsys().Rename(backup, binaryPath)
		return fmt.Errorf("verification failed: %s", strings.TrimSpace(string(out)))
	}
	_ = getFsys().Remove(backup)
	logger.Success(fmt.Sprintf("Updated paw: v%s -> v%s", currentVersion, latest))
	return nil
}

func compareVersions(a string, b string) int {
	a = strings.TrimPrefix(a, "v")
	b = strings.TrimPrefix(b, "v")
	as := strings.Split(a, ".")
	bs := strings.Split(b, ".")
	max := len(as)
	if len(bs) > max {
		max = len(bs)
	}
	for i := 0; i < max; i++ {
		var ai, bi int
		if i < len(as) {
			fmt.Sscanf(as[i], "%d", &ai)
		}
		if i < len(bs) {
			fmt.Sscanf(bs[i], "%d", &bi)
		}
		if ai < bi {
			return -1
		}
		if ai > bi {
			return 1
		}
	}
	return 0
}

func mapArch(arch string) string {
	switch arch {
	case "amd64":
		return "x64"
	case "arm64":
		return "arm64"
	default:
		return arch
	}
}

func copyFile(src string, dst string) error {
	in, err := getFsys().ReadFile(src)
	if err != nil {
		return err
	}
	return getFsys().WriteFile(dst, in, 0o755)
}

func loadState() (*updateState, error) {
	path, err := state.UpdateStatePath()
	if err != nil {
		return nil, err
	}
	data, err := getFsys().ReadFile(path)
	if err != nil {
		return nil, nil
	}
	var st updateState
	if err := json.Unmarshal(data, &st); err != nil {
		return nil, err
	}
	return &st, nil
}

func saveState(st updateState) error {
	path, err := state.UpdateStatePath()
	if err != nil {
		return err
	}
	payload, err := json.MarshalIndent(st, "", "  ")
	if err != nil {
		return err
	}
	// Fix I4: use 0o600 to restrict state file to owner-only
	return getFsys().WriteFile(path, payload, 0o600)
}
